### PR TITLE
[feature] Use jdbc connection pooling

### LINKED ISSUE #38: Databse connection pooling
It would be beneficiary to the overall performance if we had database connection pools.

### PR DESCRIPTION
**Pull request by wallner**
This pull request uses [HikariCP](https://github.com/brettwooldridge/HikariCP)
to provide JDBC connection pooling. it fixes #38

It also removes an unused dependency from the pom which doesn't really warrant
it's own pull request.

### CODE REVIEW

**Comment by tkrille** [comment]
Integration tests are failing. Here's part of a stacktrace I found in the Tomcat's logs:

```
Caused by: org.springframework.beans.BeanInstantiationException: Failed to instantiate [com.zaxxer.hikari.HikariDataSource]: Constructor threw exception; nested exception is java.lang.ClassCastException: org.postgresql.Driver cannot be cast to javax.sql.DataSource
    at org.springframework.beans.BeanUtils.instantiateClass(BeanUtils.java:163)
    at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:122)
    at org.springframework.beans.factory.support.ConstructorResolver.autowireConstructor(ConstructorResolver.java:267)
    ... 46 more
Caused by: java.lang.ClassCastException: org.postgresql.Driver cannot be cast to javax.sql.DataSource
    at com.zaxxer.hikari.pool.PoolUtilities.initializeDataSource(PoolUtilities.java:102)
    at com.zaxxer.hikari.pool.BaseHikariPool.<init>(BaseHikariPool.java:156)
    at com.zaxxer.hikari.pool.HikariPool.<init>(HikariPool.java:60)
    at com.zaxxer.hikari.pool.HikariPool.<init>(HikariPool.java:48)
    at com.zaxxer.hikari.HikariDataSource.<init>(HikariDataSource.java:80)
    at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
    at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:57)
    at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
    at java.lang.reflect.Constructor.newInstance(Constructor.java:526)
    at org.springframework.beans.BeanUtils.instantiateClass(BeanUtils.java:147)
    ... 48 more
```

I suggest you use the more traditional approach by using `jdbcUrl` for configuring Hikari.

**Comment by dacrome** [comment]
Fixes #38

**Comment by tkrille** [comment]
FYI @dacrome: [Closing issues via commit messages](https://help.github.com/articles/closing-issues-via-commit-messages/) only works in commits or in the description of a PR, but not in arbitrary comments. I've tried the same as you did, but ended up editing the description of the PR.

**Comment by wallner** [comment]
Well, it only fixes #38 when we merge it, right?

**Comment by dacrome** [comment]
Yes. Didn't see that you already mentioned this in the PR description. Anyway, GitHub only recognize this in PR desciptions or commit messages, like @tkrille already described.

**Comment by tkrille** [comment]
Can you squash the last 2 commits?

File: src/main/webapp/WEB-INF/jpa-configuration.xml [hunk 1]
```diff
@@ -50,11 +50,21 @@
...
-        <property name="url" value="${org.osiam.resource-server.db.url}"/>
-        <property name="username" value="${org.osiam.resource-server.db.username}"/>
-        <property name="password" value="${org.osiam.resource-server.db.password}"/>
+    <bean id="hikariConfig" class="com.zaxxer.hikari.HikariConfig">
+        <property name="poolName" value="springHikariCP" />
+        <property name="jdbcUrl" value="${org.osiam.resource-server.db.url}" />
+        <property name="driverClassName" value="${org.osiam.resource-server.db.driver}" />
+        <property name="dataSourceProperties">
+            <props>
+                <prop key="url">${org.osiam.resource-server.db.url}</prop>
```
Comment (tkrille) [inline src/main/webapp/WEB-INF/jpa-configuration.xml]: Is this still necessary? After all, you already specified the JDBC URL above.

Comment (wallner) [inline src/main/webapp/WEB-INF/jpa-configuration.xml]: Of course I can squash the commits, If that's preferred.

Comment (tkrille) [inline src/main/webapp/WEB-INF/jpa-configuration.xml]: I would prefer this. I would also prefer to not answer this question as a comment to a totally unrelated line comment ;)

Comment (wallner) [inline src/main/webapp/WEB-INF/jpa-configuration.xml]: Removed.

**Comment by wallner** [comment]
Squashed

File: src/main/webapp/WEB-INF/jpa-configuration.xml [hunk 2]
```diff
@@ -50,11 +50,16 @@
         </property>
     </bean>
 
-    <bean id="dataSource" class="org.springframework.jdbc.datasource.DriverManagerDataSource">
-        <property name="driverClassName" value="${org.osiam.resource-server.db.driver}"/>
-        <property name="url" value="${org.osiam.resource-server.db.url}"/>
-        <property name="username" value="${org.osiam.resource-server.db.username}"/>
-        <property name="password" value="${org.osiam.resource-server.db.password}"/>
+    <bean id="hikariConfig" class="com.zaxxer.hikari.HikariConfig">
+        <property name="poolName" value="springHikariCP" />
```
Comment (dacrome) [inline src/main/webapp/WEB-INF/jpa-configuration.xml]: I'm not sure if the pool name is logged or used somewhere (maybe later with metrics?), but using the name of an hikari spring example isn't that helpful. My suggestion: `osiam-resource-server-connection-pool`

File: pom.xml [hunk 3]
```diff
@@ -252,12 +259,6 @@
         </dependency>
 
         <dependency>
-            <groupId>commons-logging</groupId>
-            <artifactId>commons-logging</artifactId>
```
Comment (dacrome) [inline pom.xml:256]: removing this dependency is out of scope of this PR

Comment (dacrome) [inline pom.xml:256]: Ok sorry, you wrote in the PR description

> It also removes an unused dependency from the pom which doesn't really warrant it's own pull request.

but actually the pom needs to be cleanup anyway and this would justifies an own PR.

File: src/main/webapp/WEB-INF/jpa-configuration.xml (hunk 2 above)
Comment (dacrome) [inline src/main/webapp/WEB-INF/jpa-configuration.xml]: [Description of the poolName](https://github.com/brettwooldridge/HikariCP#frequently-used):

> This property represents a user-defined name for the connection pool and appears mainly in logging and JMX management consoles to identify pools and pool configurations. Default: auto-generated

Comment (tkrille) [inline src/main/webapp/WEB-INF/jpa-configuration.xml]: Yes, a better pool name would be great.

Comment (wallner) [inline src/main/webapp/WEB-INF/jpa-configuration.xml]: Done.

File: pom.xml (hunk 3 above)
Comment (tkrille) [inline pom.xml:256]: Better now than never ...

**Comment by tkrille** [comment]
LGTM now. Merged.
